pkg/virt-controller/network/ipamclaims.go

```diff
@@ -0,0 +1,195 @@
+ *
+ */
+
+package network
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/network/namescheme"
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+
+	"kubevirt.io/client-go/kubecli"
+
+	ipamclaims "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1"
+)
+
+type IPAMClaimParams struct {
+	claimName   string
+	networkName string
+}
+
+func GetNetworkToIPAMClaimParams(virtClient kubecli.KubevirtClient, namespace string, vmiName string, networks []virtv1.Network) (map[string]IPAMClaimParams, error) {
+	networkToIPAMClaimName := make(map[string]IPAMClaimParams)
+	for _, network := range vmispec.FilterMultusNonDefaultNetworks(networks) {
+		persistentIPsNetworkName, err := getPersistentIPsNetworkName(virtClient, namespace, network.Multus.NetworkName)
+		if err != nil {
+			return map[string]IPAMClaimParams{}, fmt.Errorf("failed retrieving persistentIPsNetworkName: %w", err)
+		}
+		if persistentIPsNetworkName != "" {
+			networkToIPAMClaimName[network.Name] = IPAMClaimParams{
+				claimName:   fmt.Sprintf("%s.%s", vmiName, network.Name),
+				networkName: persistentIPsNetworkName,
+			}
+		}
+	}
+
+	return networkToIPAMClaimName, nil
+}
+
+func getPersistentIPsNetworkName(virtClient kubecli.KubevirtClient, namespace string, fullNetworkName string) (string, error) {
+	namespace, networkName := getNamespaceAndNetworkName(namespace, fullNetworkName)
+	nad, err := virtClient.NetworkClient().K8sCniCncfIoV1().NetworkAttachmentDefinitions(namespace).Get(context.Background(), networkName, v1.GetOptions{})
+	if err != nil {
+		return "", fmt.Errorf("failed to locate network attachment definition %s/%s", namespace, networkName)
+	}
+
+	if nad.Spec.Config == "" {
+		return "", nil
+	}
+
+	netConf := struct {
+		AllowPersistentIPs bool   `json:"allowPersistentIPs,omitempty"`
+		Name               string `json:"name,omitempty"`
+	}{}
+	err = json.Unmarshal([]byte(nad.Spec.Config), &netConf)
+	if err != nil {
+		return "", fmt.Errorf("failed to unmarshal NAD spec.config JSON: %v", err)
+	}
+
+	if !netConf.AllowPersistentIPs {
+		return "", nil
+	}
+
+	if netConf.Name == "" {
+		return "", fmt.Errorf("failed to obtain network name: cannot be empty when allowPersistentIPs is true")
+	}
+
+	return netConf.Name, nil
+}
+
+func CreateIPAMClaims(client kubecli.KubevirtClient, vmi *virtv1.VirtualMachineInstance, ownerRef *v1.OwnerReference) error {
+	nonAbsentNetworks := filterNonAbsentNetworks(vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks)
+	networkToIPAMClaimParams, err := GetNetworkToIPAMClaimParams(client, vmi.Namespace, vmi.Name, nonAbsentNetworks)
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	// For a standalone VMI, the IPAMClaim owner is the VMI itself
+	if ownerRef == nil {
```

Comment:
IMHO it would make sense to always pass this from the client. Why should we be handling the nil scenario here ? Just ensure you call the func with this parameter always from the caller.